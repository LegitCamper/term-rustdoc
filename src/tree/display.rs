use super::*;
use std::fmt::{self, Debug};

/// skip formatting the field when the value is empty or false
macro_rules! skip_fmt {
    ($base:ident, $self:ident . $($field:ident)+ ) => {$(
        if !$self.$field.is_empty() {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
    (bool: $base:ident, $self:ident . $($field:ident)+ ) => {$(
        if $self.$field {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
    (option: $base:ident, $self:ident . $($field:ident)+ ) => {$(
        if $self.$field.is_some() {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
    (0: $base:ident, $self:ident . $($field:ident)+ ) => {$(
        if $self.$field != 0 {
            $base.field(::std::stringify!($field), &$self.$field);
        }
    )+};
}

impl Debug for DModule {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DModule");
        base.field("id", &self.id);
        skip_fmt!(
            base, self . modules structs unions enums
            functions traits constants statics type_alias imports macros
        );
        base.finish()
    }
}

impl Debug for DImpl {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DImpl");
        skip_fmt!(base, self . inherent trait_ auto blanket);
        base.finish()
    }
}

impl Debug for DStruct {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DStruct");
        base.field("id", &self.id);
        skip_fmt!(bool: base, self.contain_private_fields);
        skip_fmt!(base, self . fields impls);
        base.finish()
    }
}

impl Debug for DUnion {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DUnion");
        base.field("id", &self.id);
        skip_fmt!(
            base, self . fields impls
        );
        base.finish()
    }
}

impl Debug for DEnum {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DEnum");
        base.field("id", &self.id);
        skip_fmt!(base, self . variants impls);
        base.finish()
    }
}

impl Debug for DTrait {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DTrait");
        base.field("id", &self.id);
        skip_fmt!(base, self . types constants functions implementations);
        base.finish()
    }
}

impl Debug for DTypeAlias {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("DTypeAlias");
        base.field("id", &self.id);
        skip_fmt!(option: base, self.source_path);
        base.finish()
    }
}

impl Debug for TotolCount {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut base = f.debug_struct("TotolCount");
        skip_fmt!(
            0: base, self . modules structs unions enums functions
            traits constants statics type_alias imports
            macros_decl macros_func macros_attr macros_derv
        );
        base.finish()
    }
}
