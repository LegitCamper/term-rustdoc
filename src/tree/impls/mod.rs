mod debug;

#[macro_use]
pub mod show;
