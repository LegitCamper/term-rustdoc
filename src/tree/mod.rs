use rustdoc_types::*;
use std::collections::HashMap;

mod display;

#[allow(non_snake_case)]
mod id;
pub use id::{IDs, IdToID, SliceToIds, ID};

pub type Arr<T> = Box<[T]>;
/// Use `IDs::default()` instead of `IDs::new()` to create an empty IDs.
// pub type Vec<T> = std::sync::Arc<[T]>;
// pub type Arc<T> = std::sync::Arc<T>;

// Crate.index
pub type IndexMap = HashMap<Id, Item>;
// Crate.paths
pub type PathMap = HashMap<Id, ItemSummary>;

/// Module tree with structural items.
/// All the items only carry ids without actual data.
// NOTE: small improvement by turning all the types of fields
// from Vec to Arr after instantiation.
#[derive(Default)]
pub struct DModule {
    pub id: ID,
    // If true, this module is not part of the public API,
    // but it contains items that are re-exported as public API.
    // is_stripped: bool,
    pub modules: Vec<DModule>,
    pub structs: Vec<DStruct>,
    pub unions: Vec<DUnion>,
    pub enums: Vec<DEnum>,
    pub functions: Vec<DFunctions>,
    pub traits: Vec<DTrait>,
    pub constants: Vec<DConstant>,
    pub statics: Vec<DStatic>,
    pub type_alias: Vec<DTypeAlias>,
    pub imports: Vec<Import>,
    pub macros: Vec<DMacro>,
}

impl DModule {
    pub fn new(doc: &Crate) -> Self {
        // root module/crate name
        let (id, root) = doc
            .index
            .iter()
            .find_map(|(id, item)| {
                if item.crate_id == 0 {
                    if let ItemEnum::Module(Module {
                        is_crate: true,
                        items,
                        ..
                    }) = &item.inner
                    {
                        return Some((id.clone(), items.as_slice()));
                    }
                }
                None
            })
            .expect("root module not found");
        Self::new_inner(id.to_ID(), root, &doc.index)
    }

    fn new_inner(id: ID, inner_items: &[Id], index: &IndexMap) -> Self {
        let mut dmod = DModule {
            id,
            ..Default::default()
        };
        dmod.extract_items(inner_items, index);
        dmod
    }

    fn extract_items(&mut self, inner_items: &[Id], index: &IndexMap) {
        for item_id in inner_items {
            match index.get(item_id) {
                Some(item) => self.append(item, index),
                None => warn!("the local item {item_id:?} not found in Crate's index"),
            }
        }
    }

    fn append(&mut self, item: &Item, index: &IndexMap) {
        use ItemEnum::*;
        let id = item.id.to_ID();
        match &item.inner {
            Module(item) => self.modules.push(Self::new_inner(id, &item.items, index)),
            Struct(item) => self.structs.push(DStruct::new(id, item, index)),
            Union(item) => self.unions.push(DUnion::new(id, item, index)),
            Enum(item) => self.enums.push(DEnum::new(id, item, index)),
            Function(_) => self.functions.push(DFunctions::new(id)),
            Trait(item) => self.traits.push(DTrait::new(id, item, index)),
            Constant(_) => self.constants.push(DConstant::new(id)),
            Static(_) => self.statics.push(DStatic::new(id)),
            TypeAlias(item) => self.type_alias.push(DTypeAlias::new(id, item)),
            Macro(_) => self.macros.push(DMacro::new(id, DMacroKind::Declarative)),
            ProcMacro(proc) => self.macros.push(DMacro::new(
                id,
                match proc.kind {
                    MacroKind::Bang => DMacroKind::ProcFunction,
                    MacroKind::Attr => DMacroKind::ProcAttribute,
                    MacroKind::Derive => DMacroKind::ProcDerive,
                },
            )),
            // Primitive(_) => todo!(),
            _ => (),
        }
    }
}

#[derive(Default)]
pub struct DImpl {
    pub inherent: IDs,
    pub trait_: IDs,
    pub auto: IDs,
    pub blanket: IDs,
}
impl DImpl {
    fn new(ids: &[Id], index: &IndexMap) -> Box<Self> {
        if ids.is_empty() {
            return Default::default();
        }
        let [mut inherent, mut trait_, mut auto, mut blanket]: [Vec<ID>; 4] = Default::default();
        for Id(id) in ids {
            if id.starts_with("a:") {
                auto.push(id.to_ID());
            } else if id.starts_with("b:") {
                blanket.push(id.to_ID());
            } else {
                let id = Id(id.clone());
                if let Some(item) = index.get(&id) {
                    if let ItemEnum::Impl(impl_) = &item.inner {
                        if impl_.trait_.is_none() {
                            inherent.push(id.into_ID());
                        } else {
                            trait_.push(id.into_ID());
                        }
                    } else {
                        warn!("{id:?} in Crate's index doesn't refer to an impl item");
                    }
                } else {
                    warn!("the impl with {id:?} not found in Crate's index");
                }
            }
        }
        Box::new(DImpl {
            inherent: inherent.into(),
            trait_: trait_.into(),
            auto: auto.into(),
            blanket: blanket.into(),
        })
    }
    fn is_empty(&self) -> bool {
        self.inherent.is_empty()
            && self.trait_.is_empty()
            && self.auto.is_empty()
            && self.blanket.is_empty()
    }
}

// TODO:
// *  structs, enums, and enum variants: [non_exhaustive]
//
// [non_exhaustive]: https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute

pub struct DStruct {
    pub id: ID,
    pub fields: IDs,
    pub contain_private_fields: bool,
    pub impls: Box<DImpl>,
}
impl DStruct {
    fn new(id: ID, item: &Struct, index: &IndexMap) -> Self {
        let mut contain_private_fields = false;
        let fields = match &item.kind {
            StructKind::Unit => IDs::default(),
            StructKind::Tuple(fields) => fields
                .iter()
                .filter_map(|f| {
                    let id = f.as_ref().map(|id| id.to_ID());
                    if id.is_none() {
                        contain_private_fields = true;
                    }
                    id
                })
                .collect(),
            StructKind::Plain {
                fields,
                fields_stripped,
            } => {
                contain_private_fields = *fields_stripped;
                fields.to_ids()
            }
        };
        let impls = DImpl::new(&item.impls, index);
        DStruct {
            id,
            fields,
            contain_private_fields,
            impls,
        }
    }
}

pub struct DUnion {
    pub id: ID,
    pub fields: IDs,
    pub impls: Box<DImpl>,
}
impl DUnion {
    fn new(id: ID, item: &Union, index: &IndexMap) -> Self {
        DUnion {
            id,
            fields: item.fields.to_ids(),
            impls: DImpl::new(&item.impls, index),
        }
    }
}

pub struct DEnum {
    pub id: ID,
    pub variants: IDs,
    // variants_stripped: bool, -> Does this really make sense?
    pub impls: Box<DImpl>,
}
impl DEnum {
    fn new(id: ID, item: &Enum, index: &IndexMap) -> Self {
        DEnum {
            id,
            variants: item.variants.to_ids(),
            impls: DImpl::new(&item.impls, index),
        }
    }
}

pub struct DTrait {
    pub id: ID,
    pub types: IDs,
    pub constants: IDs,
    pub functions: IDs,
    pub implementations: IDs,
}
impl DTrait {
    fn new(id: ID, item: &Trait, index: &IndexMap) -> Self {
        let [mut types, mut constants, mut functions]: [Vec<ID>; 3] = Default::default();
        let trait_id = &id;
        for id in &item.items {
            if let Some(assoc) = index.get(id) {
                let id = id.to_ID(); // id == assoc.id
                match &assoc.inner {
                    ItemEnum::AssocType { .. } => types.push(id),
                    ItemEnum::AssocConst { .. } => constants.push(id),
                    ItemEnum::Function(_) => functions.push(id),
                    _ => warn!(
                        "`{id}` should refer to an associated item \
                         (type/constant/function) in Trait `{trait_id}`"
                    ),
                }
            } else {
                warn!("the trait item {id:?} not found in Crate's index");
            }
        }
        DTrait {
            id,
            types: types.into(),
            constants: constants.into(),
            functions: functions.into(),
            implementations: item.implementations.to_ids(),
        }
    }
}

/// generate id wrapper types for simple items
macro_rules! gen_simple_items {
    ($($name:ident),+) => {$(
        #[derive(Debug)] pub struct $name { pub id: ID, }
        impl $name { fn new(id: ID) -> Self { Self { id } } }
    )+};
}
gen_simple_items!(DFunctions, DStatic, DConstant);

pub struct DTypeAlias {
    pub id: ID,
    /// points to a source type resolved as path
    ///
    /// A type alias may point to non-path-based type though.
    pub source_path: Option<ID>,
}
impl DTypeAlias {
    pub fn new(id: ID, item: &TypeAlias) -> Self {
        Self {
            id,
            source_path: match &item.type_ {
                Type::ResolvedPath(Path { id, .. }) => Some(id.to_ID()),
                _ => None,
            },
        }
    }
}

#[derive(Debug)]
pub enum DMacroKind {
    Declarative,
    ProcFunction,
    ProcAttribute,
    ProcDerive,
}
#[derive(Debug)]
pub struct DMacro {
    pub id: ID,
    pub kind: DMacroKind,
}
impl DMacro {
    pub fn new(id: ID, kind: DMacroKind) -> Self {
        Self { id, kind }
    }
}
