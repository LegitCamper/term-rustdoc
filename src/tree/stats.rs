use super::{DMacroKind, DModule};
use std::ops::{Add, AddAssign};

#[derive(Default)]
pub struct TotolCount {
    pub modules: u32,
    pub structs: u32,
    pub unions: u32,
    pub enums: u32,
    pub functions: u32,
    pub traits: u32,
    pub constants: u32,
    pub statics: u32,
    pub type_alias: u32,
    pub imports: u32,
    pub macros_decl: u32,
    pub macros_func: u32,
    pub macros_attr: u32,
    pub macros_derv: u32,
}

impl Add for TotolCount {
    type Output = TotolCount;

    fn add(self, rhs: Self) -> Self::Output {
        TotolCount {
            modules: self.modules + rhs.modules,
            structs: self.structs + rhs.structs,
            unions: self.unions + rhs.unions,
            enums: self.enums + rhs.enums,
            functions: self.functions + rhs.functions,
            traits: self.traits + rhs.traits,
            constants: self.constants + rhs.constants,
            statics: self.statics + rhs.statics,
            type_alias: self.type_alias + rhs.type_alias,
            imports: self.imports + rhs.imports,
            macros_decl: self.macros_decl + rhs.macros_decl,
            macros_func: self.macros_func + rhs.macros_func,
            macros_attr: self.macros_attr + rhs.macros_attr,
            macros_derv: self.macros_derv + rhs.macros_derv,
        }
    }
}

impl AddAssign for TotolCount {
    fn add_assign(&mut self, rhs: Self) {
        self.modules += rhs.modules;
        self.structs += rhs.structs;
        self.unions += rhs.unions;
        self.enums += rhs.enums;
        self.functions += rhs.functions;
        self.traits += rhs.traits;
        self.constants += rhs.constants;
        self.statics += rhs.statics;
        self.type_alias += rhs.type_alias;
        self.imports += rhs.imports;
        self.macros_decl += rhs.macros_decl;
        self.macros_func += rhs.macros_func;
        self.macros_attr += rhs.macros_attr;
        self.macros_derv += rhs.macros_derv;
    }
}

impl DModule {
    /// Count the items under current module excluding the current module itself.
    #[rustfmt::skip]
    pub fn current_items_counts(&self) -> TotolCount {
        macro_rules! len {
            ($self:ident . $( $field:ident )+ ) => { $(
                let $field = $self.$field.len().try_into()
                    .expect("the count exceeds the maximum of u32");
            )+ };
        }
        len!(self . modules structs unions enums functions 
             traits constants statics type_alias imports);
        let [mut decl, mut func, mut attr, mut derv]: [u32; 4] = Default::default();
        for m in &self.macros {
            match m.kind {
                DMacroKind::Declarative => decl += 1,
                DMacroKind::ProcFunction => func += 1,
                DMacroKind::ProcAttribute => attr += 1,
                DMacroKind::ProcDerive => derv += 1,
            }
        }
        TotolCount {
            modules, structs, unions, enums, functions, 
            traits, constants, statics, type_alias, imports,
            macros_decl: decl,
            macros_func: func,
            macros_attr: attr,
            macros_derv: derv,
        }
    }

    /// Count all the items excluding the root itself.
    pub fn recursive_items_counts(&self) -> TotolCount {
        self.modules.iter().map(Self::current_items_counts).fold(
            self.current_items_counts(),
            |mut acc, tc| {
                acc += tc;
                acc
            },
        )   
    }
}
