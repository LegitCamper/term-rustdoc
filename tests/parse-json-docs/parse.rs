use crate::{doc, shot, snap};
use similar_asserts::assert_eq;
use term_rustdoc::tree::{Show, TreeLines};

#[test]
fn parse_module() {
    let (treelines, empty) = TreeLines::new_with(doc(), |doc| doc.dmodule_show_prettier());
    let doc = treelines.doc();
    let dmod = doc.dmodule();
    snap!("DModule", dmod);
    shot!("show-id", dmod.show());

    let tree = doc.dmodule_show_prettier();
    let display = tree.to_string();
    shot!("show-prettier", display);

    let display_new = treelines.display_as_plain_text();
    assert_eq!(expected: display, actual: display_new);
    snap!("flatten-tree", treelines.all_lines());
    shot!("empty-tree-with-same-depth", empty);

    // item tree
    shot!("item-tree", dmod.item_tree(&doc));

    snap!(dmod.current_items_counts(), @r###"
    ItemCount {
        modules: 1,
        structs: 2,
        functions: 3,
        traits: 1,
        constants: 2,
        macros_decl: 1,
    }
    "###);
    snap!(dmod.recursive_items_counts(), @r###"
    ItemCount {
        modules: 2,
        structs: 2,
        enums: 1,
        functions: 3,
        traits: 1,
        constants: 2,
        macros_decl: 1,
    }
    "###);

    snap!(dmod.current_impls_counts(), @r###"
    ImplCounts {
        total: ImplCount {
            kind: Both,
            total: 3,
            structs: 3,
        },
        inherent: ImplCount {
            kind: Inherent,
            total: 1,
            structs: 1,
        },
        trait: ImplCount {
            kind: Trait,
            total: 2,
            structs: 2,
        },
    }
    "###);
    snap!(dmod.recursive_impls_counts(), @r###"
    ImplCounts {
        total: ImplCount {
            kind: Both,
            total: 5,
            structs: 3,
            enums: 2,
        },
        inherent: ImplCount {
            kind: Both,
            total: 2,
            structs: 1,
            enums: 1,
        },
        trait: ImplCount {
            kind: Trait,
            total: 3,
            structs: 2,
            enums: 1,
        },
    }
    "###);

    // struct inner
    let (struct_, _) = TreeLines::new_with(treelines.doc(), |doc| {
        doc.dmodule().structs[0].show_prettier(doc)
    });
    shot!(struct_.display_as_plain_text(), @r###"
    integration::AUnitStruct
    ├── No Fields!
    └── Implementations
        ├── Trait Impls
        │   └── AUnitStruct: ATrait
        ├── Auto Impls
        │   ├── AUnitStruct: RefUnwindSafe
        │   ├── AUnitStruct: Send
        │   ├── AUnitStruct: Sync
        │   ├── AUnitStruct: Unpin
        │   └── AUnitStruct: UnwindSafe
        └── Blanket Impls
            ├── T: Any
            ├── T: Borrow<T>
            ├── T: BorrowMut<T>
            ├── T: From<T>
            ├── T: Into<U>
            ├── T: TryFrom<U>
            └── T: TryInto<U>
    "###);
}
