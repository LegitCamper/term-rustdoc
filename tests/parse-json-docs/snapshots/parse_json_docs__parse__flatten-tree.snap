---
source: tests/parse-json-docs/parse.rs
expression: treelines.all_lines()
---
[
    TreeLine {
        tag: Module,
        level: 0,
        name.text: "integration",
    },
    TreeLine {
        tag: Functions,
        level: 1,
        name.text: "Functions",
    },
    TreeLine {
        tag: Function,
        level: 2,
        name.text: "func_dyn_trait",
    },
    TreeLine {
        tag: Function,
        level: 2,
        name.text: "func_with_1arg",
    },
    TreeLine {
        tag: Function,
        level: 2,
        name.text: "func_with_1arg_and_ret",
    },
    TreeLine {
        tag: Function,
        level: 2,
        name.text: "func_with_no_args",
    },
    TreeLine {
        tag: Constants,
        level: 1,
        name.text: "Constants",
    },
    TreeLine {
        tag: Constant,
        level: 2,
        name.text: "ACONSTANT",
    },
    TreeLine {
        tag: Constant,
        level: 2,
        name.text: "ASTATIC",
    },
    TreeLine {
        tag: MacroDecls,
        level: 1,
        name.text: "Macros - Declarative",
    },
    TreeLine {
        tag: MacroDecl,
        level: 2,
        name.text: "a_decl_macro",
    },
    TreeLine {
        tag: Traits,
        level: 1,
        name.text: "Traits",
    },
    TreeLine {
        tag: Trait,
        level: 2,
        name.text: "integration::ATrait",
    },
    TreeLine {
        tag: Implementors,
        level: 3,
        name.text: "Implementors",
    },
    TreeLine {
        tag: Implementor,
        level: 4,
        name.text: "AUnitStruct: ATrait",
    },
    TreeLine {
        tag: Structs,
        level: 1,
        name.text: "Structs",
    },
    TreeLine {
        tag: Struct,
        level: 2,
        name.text: "integration::AUnitStruct",
    },
    TreeLine {
        tag: NoFields,
        level: 3,
        name.text: "No Fields!",
    },
    TreeLine {
        tag: Implementations,
        level: 3,
        name.text: "Implementations",
    },
    TreeLine {
        tag: TraitImpls,
        level: 4,
        name.text: "Trait Impls",
    },
    TreeLine {
        tag: ImplTrait,
        level: 5,
        name.text: "AUnitStruct: ATrait",
    },
    TreeLine {
        tag: AutoImpls,
        level: 4,
        name.text: "Auto Impls",
    },
    TreeLine {
        tag: ImplAuto,
        level: 5,
        name.text: "AUnitStruct: RefUnwindSafe",
    },
    TreeLine {
        tag: ImplAuto,
        level: 5,
        name.text: "AUnitStruct: Send",
    },
    TreeLine {
        tag: ImplAuto,
        level: 5,
        name.text: "AUnitStruct: Sync",
    },
    TreeLine {
        tag: ImplAuto,
        level: 5,
        name.text: "AUnitStruct: Unpin",
    },
    TreeLine {
        tag: ImplAuto,
        level: 5,
        name.text: "AUnitStruct: UnwindSafe",
    },
    TreeLine {
        tag: BlanketImpls,
        level: 4,
        name.text: "Blanket Impls",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: Any",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: Borrow<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: BorrowMut<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: From<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: Into<U>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: TryFrom<U>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: TryInto<U>",
    },
    TreeLine {
        tag: Struct,
        level: 2,
        name.text: "integration::FieldsNamedStruct",
    },
    TreeLine {
        tag: Fields,
        level: 3,
        name.text: "Fields",
    },
    TreeLine {
        tag: Field,
        level: 4,
        name.text: "field1",
    },
    TreeLine {
        tag: Field,
        level: 4,
        name.text: "field2",
    },
    TreeLine {
        tag: Field,
        level: 4,
        name.text: "field3",
    },
    TreeLine {
        tag: FieldsPrivate,
        level: 4,
        name.text: "/* private fields */",
    },
    TreeLine {
        tag: Implementations,
        level: 3,
        name.text: "Implementations",
    },
    TreeLine {
        tag: InherentImpls,
        level: 4,
        name.text: "Inherent Impls",
    },
    TreeLine {
        tag: Function,
        level: 5,
        name.text: "by_rc",
    },
    TreeLine {
        tag: Function,
        level: 5,
        name.text: "by_ref",
    },
    TreeLine {
        tag: Function,
        level: 5,
        name.text: "by_ref_mut",
    },
    TreeLine {
        tag: Function,
        level: 5,
        name.text: "consume",
    },
    TreeLine {
        tag: Function,
        level: 5,
        name.text: "new",
    },
    TreeLine {
        tag: TraitImpls,
        level: 4,
        name.text: "Trait Impls",
    },
    TreeLine {
        tag: ImplTrait,
        level: 5,
        name.text: "FieldsNamedStruct: Default",
    },
    TreeLine {
        tag: Function,
        level: 6,
        name.text: "default",
    },
    TreeLine {
        tag: AutoImpls,
        level: 4,
        name.text: "Auto Impls",
    },
    TreeLine {
        tag: ImplAuto,
        level: 5,
        name.text: "FieldsNamedStruct: RefUnwindSafe",
    },
    TreeLine {
        tag: ImplAuto,
        level: 5,
        name.text: "FieldsNamedStruct: Send",
    },
    TreeLine {
        tag: ImplAuto,
        level: 5,
        name.text: "FieldsNamedStruct: Sync",
    },
    TreeLine {
        tag: ImplAuto,
        level: 5,
        name.text: "FieldsNamedStruct: Unpin",
    },
    TreeLine {
        tag: ImplAuto,
        level: 5,
        name.text: "FieldsNamedStruct: UnwindSafe",
    },
    TreeLine {
        tag: BlanketImpls,
        level: 4,
        name.text: "Blanket Impls",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: Any",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: Borrow<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: BorrowMut<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: From<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: Into<U>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: TryFrom<U>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 5,
        name.text: "T: TryInto<U>",
    },
    TreeLine {
        tag: Module,
        level: 1,
        name.text: "integration::submod1",
    },
    TreeLine {
        tag: Enums,
        level: 2,
        name.text: "Enums",
    },
    TreeLine {
        tag: Enum,
        level: 3,
        name.text: "integration::submod1::AUnitEnum",
    },
    TreeLine {
        tag: Variants,
        level: 4,
        name.text: "Variants",
    },
    TreeLine {
        tag: Variant,
        level: 5,
        name.text: "A",
    },
    TreeLine {
        tag: Variant,
        level: 5,
        name.text: "B",
    },
    TreeLine {
        tag: Variant,
        level: 5,
        name.text: "C",
    },
    TreeLine {
        tag: Implementations,
        level: 4,
        name.text: "Implementations",
    },
    TreeLine {
        tag: InherentImpls,
        level: 5,
        name.text: "Inherent Impls",
    },
    TreeLine {
        tag: Function,
        level: 6,
        name.text: "print",
    },
    TreeLine {
        tag: TraitImpls,
        level: 5,
        name.text: "Trait Impls",
    },
    TreeLine {
        tag: ImplTrait,
        level: 6,
        name.text: "AUnitEnum: Debug",
    },
    TreeLine {
        tag: Function,
        level: 7,
        name.text: "fmt",
    },
    TreeLine {
        tag: AutoImpls,
        level: 5,
        name.text: "Auto Impls",
    },
    TreeLine {
        tag: ImplAuto,
        level: 6,
        name.text: "AUnitEnum: RefUnwindSafe",
    },
    TreeLine {
        tag: ImplAuto,
        level: 6,
        name.text: "AUnitEnum: Send",
    },
    TreeLine {
        tag: ImplAuto,
        level: 6,
        name.text: "AUnitEnum: Sync",
    },
    TreeLine {
        tag: ImplAuto,
        level: 6,
        name.text: "AUnitEnum: Unpin",
    },
    TreeLine {
        tag: ImplAuto,
        level: 6,
        name.text: "AUnitEnum: UnwindSafe",
    },
    TreeLine {
        tag: BlanketImpls,
        level: 5,
        name.text: "Blanket Impls",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 6,
        name.text: "T: Any",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 6,
        name.text: "T: Borrow<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 6,
        name.text: "T: BorrowMut<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 6,
        name.text: "T: From<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 6,
        name.text: "T: Into<U>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 6,
        name.text: "T: TryFrom<U>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 6,
        name.text: "T: TryInto<U>",
    },
    TreeLine {
        tag: Module,
        level: 2,
        name.text: "integration::submod1::submod2",
    },
    TreeLine {
        tag: Traits,
        level: 3,
        name.text: "Traits",
    },
    TreeLine {
        tag: Trait,
        level: 4,
        name.text: "integration::submod1::submod2::ATraitNeverImplementedForTypes",
    },
    TreeLine {
        tag: NoAssocOrImpls,
        level: 5,
        name.text: "No Associated Items Or Implementors!",
    },
    TreeLine {
        tag: Structs,
        level: 3,
        name.text: "Structs",
    },
    TreeLine {
        tag: Struct,
        level: 4,
        name.text: "AStructAlias",
    },
    TreeLine {
        tag: NoFields,
        level: 5,
        name.text: "No Fields!",
    },
    TreeLine {
        tag: Implementations,
        level: 5,
        name.text: "Implementations",
    },
    TreeLine {
        tag: TraitImpls,
        level: 6,
        name.text: "Trait Impls",
    },
    TreeLine {
        tag: ImplTrait,
        level: 7,
        name.text: "AUnitStruct: ATrait",
    },
    TreeLine {
        tag: AutoImpls,
        level: 6,
        name.text: "Auto Impls",
    },
    TreeLine {
        tag: ImplAuto,
        level: 7,
        name.text: "AUnitStruct: RefUnwindSafe",
    },
    TreeLine {
        tag: ImplAuto,
        level: 7,
        name.text: "AUnitStruct: Send",
    },
    TreeLine {
        tag: ImplAuto,
        level: 7,
        name.text: "AUnitStruct: Sync",
    },
    TreeLine {
        tag: ImplAuto,
        level: 7,
        name.text: "AUnitStruct: Unpin",
    },
    TreeLine {
        tag: ImplAuto,
        level: 7,
        name.text: "AUnitStruct: UnwindSafe",
    },
    TreeLine {
        tag: BlanketImpls,
        level: 6,
        name.text: "Blanket Impls",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: Any",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: Borrow<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: BorrowMut<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: From<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: Into<U>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: TryFrom<U>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: TryInto<U>",
    },
    TreeLine {
        tag: Struct,
        level: 4,
        name.text: "AUnitStruct",
    },
    TreeLine {
        tag: NoFields,
        level: 5,
        name.text: "No Fields!",
    },
    TreeLine {
        tag: Implementations,
        level: 5,
        name.text: "Implementations",
    },
    TreeLine {
        tag: TraitImpls,
        level: 6,
        name.text: "Trait Impls",
    },
    TreeLine {
        tag: ImplTrait,
        level: 7,
        name.text: "AUnitStruct: ATrait",
    },
    TreeLine {
        tag: AutoImpls,
        level: 6,
        name.text: "Auto Impls",
    },
    TreeLine {
        tag: ImplAuto,
        level: 7,
        name.text: "AUnitStruct: RefUnwindSafe",
    },
    TreeLine {
        tag: ImplAuto,
        level: 7,
        name.text: "AUnitStruct: Send",
    },
    TreeLine {
        tag: ImplAuto,
        level: 7,
        name.text: "AUnitStruct: Sync",
    },
    TreeLine {
        tag: ImplAuto,
        level: 7,
        name.text: "AUnitStruct: Unpin",
    },
    TreeLine {
        tag: ImplAuto,
        level: 7,
        name.text: "AUnitStruct: UnwindSafe",
    },
    TreeLine {
        tag: BlanketImpls,
        level: 6,
        name.text: "Blanket Impls",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: Any",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: Borrow<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: BorrowMut<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: From<T>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: Into<U>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: TryFrom<U>",
    },
    TreeLine {
        tag: ImplBlanket,
        level: 7,
        name.text: "T: TryInto<U>",
    },
]
